src/Storages/Distributed/DistributedAsyncInsertDirectoryQueue.h

```diff
@@ -142,6 +143,8 @@ class DistributedAsyncInsertDirectoryQueue
     const std::chrono::milliseconds default_sleep_time;
     std::chrono::milliseconds sleep_time;
     const std::chrono::milliseconds max_sleep_time;
+    const size_t max_retries = 0;
+    std::map<std::string, size_t> files_retry;
```

Comment:
And one elephant in the room that I missed: `files_retry` might be accessed from multiple threads, thus it has to be thread-safe. Please use a thread-safe container.

Reply:
ok，I'll add a mutex to make it thread-safe

Reply:
Ahh, sorry I was wrong. I think it is fine like this, because only pending files can be accessed from multiple places. Sorry for the inconvinience.